src/main/java/uk/gov/hmcts/divorce/divorcecase/model/Solicitor.java

```diff
@@ -94,4 +104,52 @@ public boolean hasOrgId() {
         }
         return false;
     }
+
+    public String getAddress() {
+        if ((YesOrNo.NO == getAddressOverseas() || null == getAddressOverseas()) && (this.address != null && !this.address.equals(""))) {
+            String ukPostcodeRegex = "([A-Za-z][A-Ha-hJ-Yj-y]?[0-9][A-Za-z0-9]? ?[0-9][A-Za-z]{2}|[Gg][Ii][Rr] ?0[Aa]{2})";
+            Matcher matcher = Pattern.compile(ukPostcodeRegex).matcher(this.address);
+            if (matcher.find()) {
+                String postcode =  matcher.group(1);
+                String addressPostcodeStripped = this.address.replace(postcode, "")
+                    .replace("\n\n", "\n")
+                    .replace(",,", ",")
+                    .replace("\n,\n", "\n")
+                    .replace(", , ", ", ")
+                    .replace(" \n, \n", " \n")
+                    .replace("\n ,\n", "\n")
+                    .replace(", \n ,", ",")
+                    .replace(" \n  \n", " \n")
+                    .replace("\n \n ", "\n ")
+                    .replace(" \n \n", " \n")
+                    .stripTrailing();
+
+                if (addressPostcodeStripped.contains("\n")) {
+                    String upToFirstNewLine = addressPostcodeStripped.substring(0, addressPostcodeStripped.indexOf("\n"));
+                    if (upToFirstNewLine.endsWith(",") && !addressPostcodeStripped.endsWith(",")) {
+                        addressPostcodeStripped += ",";
+                    } else if (upToFirstNewLine.endsWith(", ")
+                        && (!addressPostcodeStripped.endsWith(",") && !addressPostcodeStripped.endsWith(", "))) {
+                        addressPostcodeStripped += ", ";
+                    } else if (upToFirstNewLine.endsWith(" ") && !addressPostcodeStripped.endsWith(" ")) {
+                        addressPostcodeStripped += " ";
+                    }
+                    addressPostcodeStripped += "\n";
+                    if (addressPostcodeStripped.substring(0, upToFirstNewLine.length() + 2).endsWith(" ")) {
+                        addressPostcodeStripped += " ";
+                    }
+                } else if (this.address.substring(0, this.address.indexOf(postcode)).endsWith(",")
+                    && !addressPostcodeStripped.endsWith(",")) {
+                    addressPostcodeStripped += ",";
+                } else if (this.address.substring(0, this.address.indexOf(postcode)).endsWith(", ")
+                    && !addressPostcodeStripped.endsWith(", ")) {
+                    addressPostcodeStripped += ", ";
+                }
+                addressPostcodeStripped += this.address.endsWith("\n") ? postcode + "\n" : postcode;
+                return addressPostcodeStripped;
+            }
+        }
+
+        return this.address;
+    }
```

Comment:
Question about approach: Do we have to remove the postcode from its original position or would it be okay to leave the original address as it is and append a new line at the end with the postcode? It might look weird to repeat the postcode, but then we won't need to worry about removing useful information or breaking the formatting of the address

Reply:
I think we're going to have to move it, but I have raised the question.

Reply:
I have had a chat with Nicola and Kate.  The problem with duplicating the postcode is that we add a new line to the address.  This may mean that the duplicated postcode is below the bottom of the address window.  That will mean we don't qualify for reduced postage rates (postcode must be the last visible line).  I came up with an alternative which has been approved - for UK addresses we're going to take the address as provided up until the end of the postcode, then strip the rest from the response.  This whole method is about to get a lot tidier and smaller!